official/coin_build/man/man3/SoError.3
.TH "SoError" 3 "Sun May 28 2017" "Version 4.0.0a" "Coin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SoError \- The \fBSoError\fP class is the base class for all the error handling classes\&.
.PP
The default error handler just prints messages on the standard error output channel, but this can be overridden by client applications\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Inventor/errors/SoError\&.h>\fP
.PP
Inherited by \fBSoDebugError\fP, \fBSoMemoryError\fP, and \fBSoReadError\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "virtual \fB~SoError\fP ()"
.br
.ti -1c
.RI "const \fBSbString\fP & \fBgetDebugString\fP (void) const"
.br
.ti -1c
.RI "virtual \fBSoType\fP \fBgetTypeId\fP (void) const"
.br
.ti -1c
.RI "SbBool \fBisOfType\fP (const \fBSoType\fP type) const"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static void \fBsetHandlerCallback\fP (SoErrorCB *const func, void *const data)"
.br
.ti -1c
.RI "static SoErrorCB * \fBgetHandlerCallback\fP (void)"
.br
.ti -1c
.RI "static void * \fBgetHandlerData\fP (void)"
.br
.ti -1c
.RI "static \fBSoType\fP \fBgetClassTypeId\fP (void)"
.br
.ti -1c
.RI "static void \fBpost\fP (const char *const format,\&.\&.\&.)"
.br
.ti -1c
.RI "static \fBSbString\fP \fBgetString\fP (const \fBSoNode\fP *const node)"
.br
.ti -1c
.RI "static \fBSbString\fP \fBgetString\fP (const \fBSoPath\fP *const path)"
.br
.ti -1c
.RI "static \fBSbString\fP \fBgetString\fP (const \fBSoEngine\fP *const engine)"
.br
.ti -1c
.RI "static void \fBinitClass\fP (void)"
.br
.ti -1c
.RI "static void \fBinitClasses\fP (void)"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "virtual SoErrorCBPtr \fBgetHandler\fP (void *&data) const"
.br
.ti -1c
.RI "void \fBsetDebugString\fP (const char *const str)"
.br
.ti -1c
.RI "void \fBappendToDebugString\fP (const char *const str)"
.br
.ti -1c
.RI "void \fBhandleError\fP (void)"
.br
.in -1c
.SS "Static Protected Member Functions"

.in +1c
.ti -1c
.RI "static void \fBdefaultHandlerCB\fP (const \fBSoError\fP *error, void *userdata)"
.br
.in -1c
.SH "Detailed Description"
.PP 
The \fBSoError\fP class is the base class for all the error handling classes\&.
.PP
The default error handler just prints messages on the standard error output channel, but this can be overridden by client applications\&. 

Being able to override the default handler is useful when you want to collect error messages upon e\&.g\&. model import for later presentation of the messages to the user in any custom manner (like for instance in a GUI messagebox)\&.
.PP
The \fBSoError\fP class is not designed to be particularly useful for 'direct use'\&. Within the Coin library it is only used through its subclasses\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "SoError::~SoError ()\fC [inline]\fP, \fC [virtual]\fP"
The default destructor does nothing\&. 
.SH "Member Function Documentation"
.PP 
.SS "void SoError::setHandlerCallback (SoErrorCB *const function, void *const data)\fC [static]\fP"
This method sets the error handler callback for messages posted via this class\&.
.PP
Note that this will not override the error/debug message handler for subclasses, these will have to be overrided by calling the subclass' \fBsetHandlerCallback()\fP method\&.
.PP
\fBSee also:\fP
.RS 4
\fBdefaultHandlerCB()\fP 
.RE
.PP

.SS "SoErrorCB * SoError::getHandlerCallback (void)\fC [static]\fP"
Returns the error handler callback for messages posted via this class\&. 
.SS "void * SoError::getHandlerData (void)\fC [static]\fP"
This method returns the pointer used for passing data back to the callback handler method\&. 
.SS "const \fBSbString\fP & SoError::getDebugString (void) const"
This method returns an \fBSbString\fP containing error info from the given error instance\&. 
.SS "\fBSoType\fP SoError::getClassTypeId (void)\fC [static]\fP"
This static method returns the \fBSoType\fP for this class\&.
.PP
\fBSee also:\fP
.RS 4
\fBgetTypeId()\fP 
.RE
.PP

.SS "\fBSoType\fP SoError::getTypeId (void) const\fC [virtual]\fP"
This method returns the \fBSoType\fP of a particular object instance\&.
.PP
\fBSee also:\fP
.RS 4
\fBgetClassTypeId()\fP 
.RE
.PP

.PP
Reimplemented in \fBSoDebugError\fP, \fBSoReadError\fP, and \fBSoMemoryError\fP\&.
.SS "SbBool SoError::isOfType (const \fBSoType\fP type) const"
This method returns \fCTRUE\fP if the error instance is of - or derived from - \fItype\fP, and \fCFALSE\fP otherwise\&. 
.SS "void SoError::post (const char *const format,  \&.\&.\&.)\fC [static]\fP"
This method posts an error message\&. The \fIformat\fP string and the trailing aguments should follow the printf() standard\&. 
.SS "\fBSbString\fP SoError::getString (const \fBSoNode\fP *const node)\fC [static]\fP"
Constructs a string identifying the \fInode\fP with name (if available) and memory pointer\&. 
.SS "\fBSbString\fP SoError::getString (const \fBSoPath\fP *const path)\fC [static]\fP"
Constructs a string identifying the \fIpath\fP with name (if available) and memory pointer\&. 
.SS "\fBSbString\fP SoError::getString (const \fBSoEngine\fP *const engine)\fC [static]\fP"
Constructs a string identifying the \fIengine\fP with name (if available) and memory pointer\&. 
.SS "void SoError::initClass (void)\fC [static]\fP"
This method takes care of initializing all static data for the class\&. 
.SS "void SoError::initClasses (void)\fC [static]\fP"
This static method initializes all the \fBSoError\fP classes\&. 
.SS "void SoError::defaultHandlerCB (const \fBSoError\fP * error, void * data)\fC [static]\fP, \fC [protected]\fP"
Contains the default code for handling error strings\&.
.PP
Default treatment of an error message is to print it out on the standard error file handle\&. 
.SS "SoErrorCB * SoError::getHandler (void *& data) const\fC [protected]\fP, \fC [virtual]\fP"
This is just a convenience wrapper around the \fBgetHandlerCallback()\fP and \fBgetHandlerData()\fP methods\&. 
.PP
Reimplemented in \fBSoDebugError\fP, \fBSoReadError\fP, and \fBSoMemoryError\fP\&.
.SS "void SoError::setDebugString (const char *const str)\fC [protected]\fP"
Replace the latest stored debug string with \fIstr\fP\&. 
.SS "void SoError::appendToDebugString (const char *const str)\fC [protected]\fP"
Add \fIstr\fP at the end of the currently stored debug string\&. 
.SS "void SoError::handleError (void)\fC [protected]\fP"
This method calls the appropriate handler for an error instance\&. All error handling goes through this method, and is therefore a good candidate for a debugger breakpoint\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for Coin from the source code\&.


insanity/coin/src/errors/SoError.cpp
/**************************************************************************\
 * Copyright (c) Kongsberg Oil & Gas Technologies AS
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\**************************************************************************/

/*!
  \class SoError SoError.h Inventor/errors/SoError.h
  \brief The SoError class is the base class for all the error handling classes.
  \ingroup errors

  The default error handler just prints messages on the standard error
  output channel, but this can be overridden by client applications.

  Being able to override the default handler is useful when you want
  to collect error messages upon e.g. model import for later
  presentation of the messages to the user in any custom manner (like
  for instance in a GUI messagebox).

  The SoError class is not designed to be particularly useful for
  "direct use". Within the Coin library it is only used through its
  subclasses.

*/

// *************************************************************************

#include <Inventor/errors/SoErrors.h>

#include "coindefs.h"

#include <cassert>
#include <cstdarg>
#include <cstdio>
#include <cstdlib>

#include <Inventor/C/tidbits.h>
#include <Inventor/SbName.h>
#include <Inventor/SoPath.h>
#include <Inventor/engines/SoEngine.h>
#include <Inventor/nodes/SoNode.h>

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif // HAVE_CONFIG_H

// *************************************************************************

SoType SoError::classTypeId STATIC_SOTYPE_INIT;
SoErrorCB * SoError::callback = nullptr; // make use of default cc_error handler
void * SoError::callbackData = nullptr;

// *************************************************************************

// "Converter" constructor.
SoError::SoError(const cc_error * error)
{
  cc_error_init(&this->err);
  cc_error_copy(error, &this->err);
}

void
SoError::callbackForwarder(const cc_error * error, void * COIN_UNUSED_ARG(data))
{
  SoError wrappederr(error);

  assert(SoError::callback != nullptr);
  (*SoError::callback)(&wrappederr, SoError::callbackData);
}


/*!
  \fn SoError::~SoError()

  The default destructor does nothing.
*/


/*!
  This method takes care of initializing all static data for the class.
*/
void
SoError::initClass(void)
{
  SoError::callback = defaultHandlerCB;
  SoError::callbackData = nullptr;
  SoError::classTypeId =
    SoType::createType(SoType::badType(), SbName("Error"));
}

/*!
  This static method initializes all the SoError classes.
*/
void
SoError::initClasses(void)
{
  SoError::initClass();
  SoDebugError::initClass();
  SoMemoryError::initClass();
  SoReadError::initClass();
}

/*!
  This static method returns the SoType for this class.

  \sa getTypeId()
*/
SoType
SoError::getClassTypeId(void)
{
  return SoError::classTypeId;
}

// FIXME: grab better version of getTypeId() doc from SoBase, SoAction
// and / or SoDetail. 20010913 mortene.
/*!
  This method returns the SoType of a particular object instance.

  \sa getClassTypeId()
*/
SoType
SoError::getTypeId(void) const
{
  return SoError::classTypeId;
}

/*!
  This method returns \c true if the error instance is of - or derived
  from - \a type, and \c false otherwise.
*/
bool
SoError::isOfType(const SoType type) const
{
  const SoType myType = this->getTypeId();
  return (myType == type) ? true : myType.isDerivedFrom(type);
}

/*!
  This method sets the error handler callback for messages posted via
  this class.

  Note that this will not override the error/debug message handler for
  subclasses, these will have to be overrided by calling the subclass'
  setHandlerCallback() method.

  \sa defaultHandlerCB()
*/
void
SoError::setHandlerCallback(SoErrorCB * const function, void * const data)
{
  if (SoError::callback == SoError::defaultHandlerCB ||
      SoError::callback == nullptr) {
    // The user is overriding the default handler, so set up a
    // "converter" callback function that makes an SoError out of an
    // cc_error and forwards control to the callback function given as
    // an argument to setHandlerCallback().
    cc_error_set_handler_callback(
       reinterpret_cast<cc_error_cb *>(SoError::callbackForwarder),
       nullptr);
  }

  SoError::callback = function;
  SoError::callbackData = data;
}

/*!
  Returns the error handler callback for messages posted via this
  class.
*/
SoErrorCB *
SoError::getHandlerCallback(void)
{
  return SoError::callback;
}

/*!
  This method returns the pointer used for passing data back to the
  callback handler method.
*/
void *
SoError::getHandlerData(void)
{
  return SoError::callbackData;
}

/*!
  This method returns an SbString containing error info from the given
  error instance.
*/
const SbString &
SoError::getDebugString(void) const
{
  // Cast away constness and fetch value from underlying cc_error instance.
  const_cast<SbString &>(this->debugstring) = cc_string_get_text(cc_error_get_debug_string(&this->err));
  return this->debugstring;
}


/*!
  This method posts an error message.  The \a format string and the
  trailing aguments should follow the printf() standard.
*/
void
SoError::post(const char * const format, ...)
{
  va_list args;
  va_start(args, format);
  cc_error_post_arglist(format, args);
  va_end(args);
}

/*!
  Constructs a string identifying the \a node with name (if available)
  and memory pointer.
*/
SbString
SoError::getString(const SoNode * const node)
{
  SbString str;
  SoError::generateBaseString(str, node, "node");
  return str;
}

/*!
  Constructs a string identifying the \a path with name (if available)
  and memory pointer.
*/
SbString
SoError::getString(const SoPath * const path)
{
  SbString str;
  SoError::generateBaseString(str, path, "path");
  return str;
}

/*!
  Constructs a string identifying the \a engine with name (if available)
  and memory pointer.
*/
SbString
SoError::getString(const SoEngine * const engine)
{
  SbString str;
  SoError::generateBaseString(str, engine, "engine");
  return str;
}

/*!
  Contains the default code for handling error strings.

  Default treatment of an error message is to print it out on the
  standard error file handle.
*/
void
SoError::defaultHandlerCB(const SoError * error, void * data)
{
  cc_error_default_handler_cb(&error->err, data);
}

/*!
  This is just a convenience wrapper around the getHandlerCallback() and
  getHandlerData() methods.
*/
SoErrorCB *
SoError::getHandler(void * & data) const
{
  data = SoError::callbackData;
  return SoError::callback;
}

/*!
  Replace the latest stored debug string with \a str.
*/
void
SoError::setDebugString(const char * const str)
{
  cc_error_set_debug_string(&this->err, str);
}

/*!
  Add \a str at the end of the currently stored debug string.
*/
void
SoError::appendToDebugString(const char * const str)
{
  cc_error_append_to_debug_string(&this->err, str);
}

/*!
  This method calls the appropriate handler for an error instance.  All
  error handling goes through this method, and is therefore a good
  candidate for a debugger breakpoint.
*/
void
SoError::handleError(void)
{
  // If this assert() statement were present earlier, it could have
  // saved me half the day of debugging today.. 20021015 mortene.
  assert((SoError::classTypeId != SoType::badType()) &&
         "SoError attempted used before class was initialized");

  cc_error_handle(&this->err);
}

/*!
  This method is used by the getString methods.  It just generates a
  '<what> named "<name>" at address <address>' string.
*/
void
SoError::generateBaseString(SbString & str,
                            const SoBase * const base,
                            const char * const what)
{
  str.sprintf("%s named \"%s\" at address %p",
              what, base->getName().getString(), base);
}
